mono/metadata/icall.c

```diff
@@ -3548,18 +3548,20 @@ ves_icall_InternalInvoke (MonoReflectionMethod *method, MonoObject *this_arg, Mo
 		}
 	}
 
-	if (sig->ret->byref) {
 #if ENABLE_NETCORE
+	if (sig->ret->byref || m_class_is_byreflike (mono_class_from_mono_type_internal (sig->ret))) {
 		MonoType* ret_byval = m_class_get_byval_arg (mono_class_from_mono_type_internal (sig->ret));
-		if (ret_byval->byref) {
+		if (ret_byval->byref || m_class_is_byreflike (mono_class_from_mono_type_internal (ret_byval))) {
```

Comment:
I'm a little bit confused about #13901 and https://github.com/dotnet/coreclr/pull/17639 - seems like the intention in CoreCLR is to forbid things like `ref Span<int>` (ByRef to ByRefLike as they say).

But we seem to be forbidding a wider class of things - for example it looks like we'll disallow `Span<int>`.

Reply:
I don't think we disallow `Span<int>` unless I messed up the check. The idea is to box one ByRef/IsByRefLike but disallow any nested ones that would require more than one boxing. The failing tests were doing `Span<ref X>`.

Reply:
@filipnavara ah... I think `Span<ref int>` should be ruled out by disallowing instantiations with ByRef types...

in the case of `Span<int>`, `sig->ret->byref` is `False` and `m_class_is_byreflike (mono_class_from_mono_type_internal (sig->ret)))` is `True`.
But then `ret_byval` is `Span<int>` again (`m_class_get_byval_arg` is just the `Span<int>` again - not `int`,  I think like you expected).

So here's a sample:
<details>
<pre>
using System;

public ref struct Q {
	public int i;
}

public struct S {
	public int j;
}

public class Foo {
	static S s;

	public static void Main () {
		
		bool threw;

		threw = false;
		try {
			var mi1 = typeof(Foo).GetMethod ("Test1");
			var o1 = mi1.Invoke (null, null);
		} catch (NotSupportedException) { threw = true; }
		Console.WriteLine ($"mi1.Invoke: threw {threw} (expect False)");
			
		threw = false;
		try {
			var mi2 = typeof(Foo).GetMethod ("Test2");
			var o2 = mi2.Invoke (null, null);
		}  catch (NotSupportedException) { threw = true; }
		Console.WriteLine ($"mi1.Invoke: threw {threw} (expect False)");


		threw = false;
		try {
			var mi3 = typeof(Foo).GetMethod ("Test3");
			var o3 = mi3.Invoke (null, null);
		} catch (NotSupportedException) { threw = true; }
		Console.WriteLine ($"mi3.Invoke: threw {threw} (expect False)");

	}

	public static Q Test1 ()
	{
		Q q;
		q.i = 42;
		Console.WriteLine ("a");
		return q;
	}

	public static Span<int> Test2 ()
	{
		int[] a = new int[] {1, 4, 9, 16};
		return new Span<int>(a);
	}

	public static ref S Test3 ()
	{
		s.j = 196;
		return ref s;
	}

}
</pre>
</details>

It ends up printing:
```
mi1.Invoke: threw True (expect False)
mi1.Invoke: threw True (expect False)
mi3.Invoke: threw False (expect False)
```

Reply:
For completeness, here's the same sample as https://github.com/mono/mono/pull/14967#discussion_r292610023 running on .NET Framework 4.8 (using VS2019 you need to add a reference to the System.Memory nuget to get a copy of `Span<T>`; but ref structs as a language feature are there already):

```
a
mi1.Invoke: threw False (expect True)
mi2.Invoke: threw False (expect True)
mi3.Invoke: threw True (expect False)
```

So... yea... at least we know that `#if ENABLE_NETCORE` is a crucial piece of the implementation here.